File: src/collectionview.ts
"use strict";

import * as _ from "underscore";
import * as Handlebars from "handlebars";

export class ACollectionView {
	public wasm_module: any = null;

	/**
	 * Конструктор представления.
	 * 
	 * @method constructor
	 * @param {any} wasm_context Контекст.
	 * @param {any} wasm_parent Родительский узел.
	 * @param {string} nodeId Уникальный идентификатор узла.
	 * @param {any} nodeCreateInfo Свойства представления.
	 */
	constructor(public wasm_context: any = null, public wasm_parent: any = null, nodeId: string, nodeCreateInfo?: any) {
		this.wasm_module = function (self: any) {
			var AItemCollectionViewObject = wasm_context.module.ACollectionView.extend("ACollectionView", {
				__construct: function(parent: any, nodeId: string, nodeCreateInfo: any) {
					this.__parent.__construct.call(this, parent, nodeId, nodeCreateInfo);
				},
				__destruct: function() {
					this.__parent.__destruct.call(this);
				},
				initialize: _.bind(self.initialize, self),
				update: _.bind(self.update, self)
			});

			return new AItemCollectionViewObject(wasm_parent, nodeId, nodeCreateInfo);
		}(this) as any;
	}

	public initialize(): void {
		// Empty
	}

	public update(): void {
		// Empty
	}
};

File: src/collectionview.cpp
#include <sway/webcore/mvc/collectionview.h>
#include <sway/webcore/mvc/itemview.h>
#include <sway/webcore/mvc/itemmodel.h>
#include <sway/webcore/treeupdater.h>

NAMESPACE_BEGIN(sway)
NAMESPACE_BEGIN(webcore)
NAMESPACE_BEGIN(mvc)

void ACollectionView::registerEmscriptenClass(lpcstr_t classname) {
	emscripten::class_<ACollectionView, emscripten::base<AVisualComponent>>(classname)
		.allow_subclass<ACollectionViewWrapper>("ACollectionViewWrapper", emscripten::constructor<core::containers::HierarchyNodePtr_t, std::string, TreeNodeElementCreateInfo>())
		.constructor<core::containers::HierarchyNodePtr_t, std::string, TreeNodeElementCreateInfo>()
		.function("initialize", emscripten::optional_override([](ACollectionView & self) {
			return self.ACollectionView::initialize();
		}))
		.function("update", emscripten::optional_override([](ACollectionView & self) {
			return self.ACollectionView::update();
		}))
		.function("makeItem", &ACollectionView::makeItem, emscripten::allow_raw_pointers());
}

ACollectionView::ACollectionView(core::containers::HierarchyNodePtr_t parent,
	//const core::containers::HierarchyNodeIndex & nodeIndex,
	const std::string & nodeId, const TreeNodeElementCreateInfo & createInfo)
	: AVisualComponent(parent, core::containers::HierarchyNodeIndex(), nodeId, createInfo) {
	// Empty
}

void ACollectionView::initialize() {
	// Empty
}

void ACollectionView::update() {
	// Empty
}

void ACollectionView::makeItem(u32_t index, TreeNodeElement * child) {
	auto collection = static_cast<ACollectionModel *>(getModel());
	auto item = static_cast<AItemView *>(child);
	item->setNodeId(core::misc::format("%s_%i", item->getNodeId().c_str(), index));
	item->setModel(collection->getItem(index));
	item->initialize();
	addChild(item);
}

void ACollectionView::accept(ITreeVisitor * visitor) {
	visitor->visitOnEnter(this);

	for (core::containers::HierarchyNodePtr_t node : getChildren())
		static_cast<ACollectionView *>(node)->accept(visitor);
}

void ACollectionViewWrapper::initialize() {
	return call<void>("initialize");
}

void ACollectionViewWrapper::update() {
	return call<void>("update");
}

NAMESPACE_END(mvc)
NAMESPACE_END(webcore)
NAMESPACE_END(sway)
